src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/stdx/optional.hpp

```diff
@@ -14,93 +14,677 @@
 
 #pragma once
 
-#include <bsoncxx/config/prelude.hpp>
+#include <cstdio>
+#include <exception>
+#include <initializer_list>
+#include <memory>
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+
+#include "./operators.hpp"
+#include "./type_traits.hpp"
 
-#if defined(BSONCXX_POLY_USE_MNMLSTC)
+// Try to detect std::optional. We'll reuse some components from there for compatibility
+// with C++17 users.
+#pragma push_macro("BSONCXX_HAVE_STDLIB_OPTIONAL")
+#undef BSONCXX_HAVE_STDLIB_OPTIONAL
+#define BSONCXX_HAVE_STDLIB_OPTIONAL 0
 
-#include <core/optional.hpp>
+#ifdef __has_include
+#if __has_include(<version>)
+#include <version>
+#if defined(__cpp_lib_optional) && __cpp_lib_optional >= 201606L
+#undef BSONCXX_HAVE_STDLIB_OPTIONAL
+#define BSONCXX_HAVE_STDLIB_OPTIONAL 1
+#include <optional>
+#endif
+#endif
+#endif
+
+#include <bsoncxx/config/prelude.hpp>
 
 namespace bsoncxx {
+
 namespace v_noabi {
+
 namespace stdx {
 
-using ::core::make_optional;
-using ::core::nullopt;
-using ::core::nullopt_t;
-using ::core::optional;
+/**
+ * @brief Implementation of an std::optional-like class template
+ *
+ * Presents mostly the same interface as std::optional from C++17.
+ *
+ * @tparam T The type being made "optional"
+ */
+template <typename T>
+class optional;
 
-}  // namespace stdx
-}  // namespace v_noabi
-}  // namespace bsoncxx
+#if BSONCXX_HAVE_STDLIB_OPTIONAL
+// If we have access to the stdlib component, re-use those so that users can use
+// the same types and constants defined in <optional>
+using std::bad_optional_access;
+using std::in_place;
+using std::in_place_t;
+using std::nullopt;
+using std::nullopt_t;
+#else
+/**
+ * @brief Exception type thrown upon attempted access to a value-less optional<T>
+ * via a throwing accessor API.
+ */
+class bad_optional_access : public std::exception {
+   public:
+    const char* what() const noexcept override {
+        return "bad_optional_access()";
+    }
+};
+/// Tag type to represent an empty optional value
+struct nullopt_t {};
+/// Tag constant to construct or compare with an empty optional value
+static constexpr nullopt_t nullopt{};
+/// Tag used to call the emplacement-constructor of optional<T>
+static constexpr struct in_place_t {
+} in_place;
+#endif
 
-#elif defined(BSONCXX_POLY_USE_BOOST)
+namespace detail {
 
-#include <boost/none.hpp>
-#include <boost/optional/optional.hpp>
-#include <boost/optional/optional_io.hpp>
+// Terminates the program when an illegal use of optional<T> is attempted
+[[noreturn]] inline void terminate_disengaged_optional(const char* what) noexcept {
+    (void)std::fprintf(stderr, "%s: Invalid attempted use of disengaged optional<T>\n", what);
+    std::terminate();
+}
```

Comment:
I believe it is fully conforming to call `terminate` in the disengaged-`operator*()`, since it is otherwise UB.

Reply:
Not that it's much justification, but it appears to be extremely easy to "accidentally" `#include <cstdio>`, at least with GCC which pulls in half the universe with `<string>`, so I think it extremely likely that most of our headers probably pull `<cstdio>` unintentionally on many platforms. Maybe someday C++ Modules will save us, but probably not in this codebase.

With that in mind, _I_ think it's an extremely useful assertion, enuogh to warrant pulling in `<cstdio>`. Alternatively, it may be useful to have a single once-ever-defined ABI-stable bsoncxx "assertion failure" handler that does the I/O. I think I'd rather go that route, but I don't know what we might want to do in the interim.